### ISSUE TITLE
Creating a relation result the foreign key to be negative

### ISSUE THREAD

**Issue by sam2x** [maritz/nohm#91]
Hi,

I find out your project who looks promising. So thanks you for providing us this ORM.
I stuck on a weirdz bug since 4hours, it's about relation.

## Context 

I have 3 models, let's name them ModelA, ModelB, ModelC. In my code i use the following relation scheme in pseudo code:

modelA = Factory('ModelA')
modelA.link(modelB)
modelC.link(modelA)
modelC.save()
modelA.save()

modelA is a new created Model, modelB & modelC has already been created, so in this context modelB & modelC are loaded(using findandload for modelB, using load for modelC).

## Log  
 
I used Redis monitor to log information. For this issue, i used simple model name. So here, 
ModelA is Shop, ModelB is Location, and ModelC is User:

>1423885279.079308 [1 127.0.0.1:49409] "get" "sess:IXRhx2QRV6sevTz7LC-e6TbLQTn25ICg"
   1423885279.080167 [0 127.0.0.1:49405] "hgetall" "nohm:\hash:User:1"
   1423885279.082581 [0 127.0.0.1:49405] "sinter" "nohm:index:Location:country:FRA"
   1423885279.082952 [0 127.0.0.1:49405] "hgetall" "nohm:\hash:Location:1"
   1423885279.083699 [0 127.0.0.1:49405] "sismember" "nohm:idsets:User" "1"
   1423885279.083852 [0 127.0.0.1:49405] "sismember" "nohm:idsets:Shop" "-1423884760366"
   1423885279.084342 [0 127.0.0.1:49405] "MULTI"
   1423885279.084389 [0 127.0.0.1:49405] "EXEC" 
   1423885279.084709 [0 127.0.0.1:49405] "incr" "nohm:ids:Shop"
   1423885279.084914 [0 127.0.0.1:49405] "MULTI"

Note: i put an \ before the "hash" string on the two hgetall commands cause emoji-smiley were displaying and i don't know how to disable this shit from github... 
## Assumption 

Not sure what's happenings, but Nohm seems to set my Shop ID to  -1423884760366 which looks like a timestamp, but i never touched/assigned the id value of Model...

I'm totally lost. May Nohm set this value by error during these two relations ? 

Regards,
SM.

**Comment by maritz** [comment]
This should be normal behaviour. In the current implementation there is a step before validation and saving that makes this negative timestamp ID if the model is new. This is required to set unique locks during validation.

What should then happen during the save is that these ids get changed to the correct ones after validation and saving. Is that not the case?

Can you provide a small test case (preferably working code that is as small as possible while producing this problem) so I can write a unit test and then fix this? Or alternatively if you find the issue yourself, feel free to write those yourself and create a pull request. :-)

**Comment by sam2x** [comment]
Hi Peter,

Thanks you for your quick answer. After some analyze, I prepared a correct small test case which reproduce the problem! ;)

http://pastebin.com/BVa9Qqu1

Here how to use 

    dev@box:~/Desktop/nohm$ node testcase_negativeRelationId.js 
    Cool, default database data has been saved! comment step 1, and uncomment step2!
    dev@box:~/Desktop/nohm$ node testcase_negativeRelationId.js 
    Cool, relations has been created! comment step 2, and uncomment step3!
    dev@box:~/Desktop/nohm$ node testcase_negativeRelationId.js 
    shopIds found (you see the negative value omg :'( ) [ -1423916970013 ]

**Comment by sam2x** [comment]
I tried to change the generator (using the default one). And this issue happens too. I'll try to dive in code, any pointers to start with ?

**Comment by sam2x** [comment]
Ok i identified the source of my problem.

On my code i had:

    // create shop
    myshop = nohm.factory('Shop', {
                 name: shopName,
    })
    myshop.link(myshoploc)
    user.link(myshop, "creator")
    user.save()
    myshop.save(function (err, data){
                if (!err){
                   console.log("Cool, relations has been created! comment step 2, and uncomment step3!")
                }
    })

The problem is i'm creating link *before* saving the new Model myshop. The solution is to logically link the model once the new model is saved : 

     // create shop
    myshop = nohm.factory('Shop', {
                 name: shopName,
    })
    myshop.link(myshoploc)
    myshop.save(function (err, data){
                if (!err){
                   user.link(myshop, "creator")
                   user.save()
                   console.log("Cool, relations has been created! comment step 2, and uncomment step3!")
                }
    })

Is this supposed to be normal ? Can we clarify the doc with this nasty testcase about thing to avoid ? 
Thanks you for your comment, it helped a lot to find out this issue.

**Comment by maritz** [comment]
No, there is something wrong and needs to be fixed.

But I'm glad that you found a workaround for now. :+1:

**Comment by sam2x** [comment]
After reviewing this thread, i suggest to return an error because the linking refer to an unsaved instance (eg error return:'model instance X linked is not saved').

Code to fix should be during save chain, but it's imho low priority.  

ps: Is this possible to have tag 'low/medium/high' in addition to 'bug' label ?

**Comment by maritz** [comment]
After re-reading it all and thinking about it a bit more, this is both a bug in your usage and a problem in nohm.

So the reason it produces a bad db state is that you don't wait for the user.save() before starting the myShop.save().

When saving a model that has unsaved other models linked, those unsaved other models get saved first. This creates a race-condition in your case where the myShop is saved twice at the same time.

In fact you can just save the user and that should cascade down to saving myShop automatically.

I'm not quite sure how to properly handle this yet. Maybe just a "per-instance" .locked property during save and then throw if it's set?

There are also potentially race-condition problems during update .save() calls that are a bit more obscure and difficult to prevent.
